reactive/apache_bigtop_base.py

```diff
@@ -1,11 +1,24 @@
-from charms.reactive import when, when_not, set_state
-from charmhelpers.core import hookenv
+from path import Path
+from charms.reactive import when, when_any, when_not, when_none
+from charms.reactive import RelationBase, set_state
+from charms.reactive.helpers import data_changed, any_states
+from charmhelpers.core import hookenv, unitdata
 from charmhelpers.core.host import ChecksumError
 from charmhelpers.fetch import UnhandledSource
 from charms.layer.apache_bigtop_base import Bigtop
 
 
 @when('puppet.available')
+@when_none('java.ready', 'hadoop-plugin.java.ready')
+def missing_java():
+    if any_states('java.joined', 'hadoop-plugin.joined'):
+        hookenv.status_set('waiting', 'Waiting on Java')
+    else:
+        hookenv.status_set('blocked', 'Waiting on relation to Java')
+
+
+@when('puppet.available')
+@when_any('java.ready', 'hadoop-plugin.java.ready')
 @when_not('bigtop.available')
 def fetch_bigtop():
```

Comment:
We get an exception here: JavaProvides is passed as parameter

unit-resourcemgr-0[2549]: 2016-05-17 11:14:04 INFO unit.resourcemgr/0.juju-log server.go:268 java:6: Invoking reactive handler: reactive/apache_bigtop_base.py:20:fetch_bigtop
unit-resourcemgr-0[2549]: 2016-05-17 11:14:04 INFO unit.resourcemgr/0.java-relation-changed logger.go:40 Traceback (most recent call last):
unit-resourcemgr-0[2549]: 2016-05-17 11:14:04 INFO unit.resourcemgr/0.java-relation-changed logger.go:40   File "/var/lib/juju/agents/unit-resourcemgr-0/charm/hooks/java-relation-changed", line 19, in <module>
unit-resourcemgr-0[2549]: 2016-05-17 11:14:04 INFO unit.resourcemgr/0.java-relation-changed logger.go:40     main()
unit-resourcemgr-0[2549]: 2016-05-17 11:14:04 INFO unit.resourcemgr/0.java-relation-changed logger.go:40   File "/usr/local/lib/python3.4/dist-packages/charms/reactive/__init__.py", line 69, in main
unit-resourcemgr-0[2549]: 2016-05-17 11:14:04 INFO unit.resourcemgr/0.java-relation-changed logger.go:40     bus.dispatch()
unit-resourcemgr-0[2549]: 2016-05-17 11:14:04 INFO unit.resourcemgr/0.java-relation-changed logger.go:40   File "/usr/local/lib/python3.4/dist-packages/charms/reactive/bus.py", line 418, in dispatch
unit-resourcemgr-0[2549]: 2016-05-17 11:14:04 INFO unit.resourcemgr/0.java-relation-changed logger.go:40     _invoke(other_handlers)
unit-resourcemgr-0[2549]: 2016-05-17 11:14:04 INFO unit.resourcemgr/0.java-relation-changed logger.go:40   File "/usr/local/lib/python3.4/dist-packages/charms/reactive/bus.py", line 401, in _invoke
unit-resourcemgr-0[2549]: 2016-05-17 11:14:04 INFO unit.resourcemgr/0.java-relation-changed logger.go:40     handler.invoke()
unit-resourcemgr-0[2549]: 2016-05-17 11:14:04 INFO unit.resourcemgr/0.java-relation-changed logger.go:40   File "/usr/local/lib/python3.4/dist-packages/charms/reactive/bus.py", line 277, in invoke
unit-resourcemgr-0[2549]: 2016-05-17 11:14:04 INFO unit.resourcemgr/0.java-relation-changed logger.go:40     self._action(*args)
unit-resourcemgr-0[2549]: 2016-05-17 11:14:04 INFO unit.resourcemgr/0.java-relation-changed logger.go:40 TypeError: fetch_bigtop() takes 0 positional arguments but 1 was given
unit-resourcemgr-0[2549]: 2016-05-17 11:14:04 ERROR juju.worker.uniter.operation runhook.go:107 hook "java-relation-changed" failed: exit status 1